### DISCUSSION TITLE
Qute: How to dynamically resolve message keys?

### DISCUSSION THREAD

**Discussion by Brutus5000** [quarkusio/quarkus#34236] [Q&A] [answered]
I'm looking for a solution to dynamically resolve message keys from a resource bundle.

What I found so far is issue #11567 pointing to the solution at https://github.com/quarkusio/quarkus/pull/11613/files#diff-c2466ea58430f533cbe0c33ca9ade344beaef598a94cb34b9dd69d7757205469

So the reference states that it should be possible to resolve runtime keys via `{msg:message('key','variables...')}`

So my setup now is:

1) an empty MessageBundle class (otherwise `msg` doesn't resolve at all in the template):
```kotlin
@MessageBundle
interface AppMessages {
}
```

2) a simple msg_en.properties file:
```properties
header_loggedInAs=You are logged in as: {name}
```

3) a template referencing this key:
```html
{msg:message('header_loggedInAs', 'brutus5000')
```

But when starting the app I get an ugly stacktrace:

```
Caused by: io.quarkus.qute.deployment.MessageBundleException: Message bundle method header_loggedInAs() not found on: com.faforever.config.AppMessages
	- file: /[...]/build/resources/main/messages/msg_en.properties
	- line 0
```

What am I doing wrong? (I'm on the latest quarkus version 3.1.2.Final)

[a comment by an automated account is not shown]

**Reply by mkouba** [reply]
You cannot use an empty message bundle interface.  You will need to add a method for each key in the properties file, e.g. something like:
```java
@MessageBundle
interface AppMessages {
   
   @Message
   String header_loggedInAs(String name);
}
```

See also [Bundle Name and Message Keys](https://quarkus.io/guides/qute-reference#bundle-name-and-message-keys) to see how the method names can be mapped to keys. 

> a template referencing this key:

If you don't need to obtain the key dynamically then you can also use something like: `{msg:header_loggedInAs('brutus5000')}`

**Reply by Brutus5000** [reply]
Yeah the hard coded message key was just a dummy test. I have cases where I get the message key from externally (OAuth2 Scopes from external provider). 
These can't be resolved at compile time. I guess rendering at runtime will crash if a message key does not exist? I'd rather have an opportunity to give a fallback value instead.

Otherwise I could run my own i18n bean I guess... just trying to figure out what can be done with Qute builtins.

**Reply by mkouba** [reply, marked as answer]
Yep, qute type-safe message bundles are not designed to handle external sources of messages, i.e. messages that cannot be validated at build time.

**Reply by Brutus5000** [reply]
But for clarification: are there type-**un**safe message bundles provided by Qute (or Quarkus) that could be used?

**Reply by mkouba** [reply]
No such type-**un**safe message bundles exist in qute/quarkus. You would need a custom solution for external message sources. Also I would recommend you to explore the [Namespace Template Extension Methods](https://quarkus.io/guides/qute-reference#namespace_extension_methods) for qute integration.

**Comment by Brutus5000** [comment]
Thank you very much.

**Reply by mkouba** [reply]
You're welcome! :smiley:
